### DISCUSSION TITLE
Bulk FHIR post-processing response lifecycle?

### DISCUSSION THREAD

**Discussion by jc00ke** [medplum/medplum#4921] [Q&A]
Hello! My question is around bulk FHIR: If I give you 100 FHIR resources in a file and 3 of them are invalid, how do I get notified of the 3 invalid resources? Is the response bespoke? I would hope not, but I haven't found anything in any of the FHIR docs, though maybe I've missed it.

Thanks!

**Comment by jc00ke** [comment]
Maybe I'm about to answer my own question...

- [Asynchronous Bulk Data Request Pattern](https://build.fhir.org/async-bulk.html)
- [Asynchronous Interaction Request Pattern](https://build.fhir.org/async-bundle.html)

**Reply by jc00ke** [reply]
![image](https://github.com/user-attachments/assets/b77441b0-57bf-4040-81c7-01af8af63425)

[From Athena](https://docs.athenahealth.com/api/workflows/fhir-bulk-export-workflow#1_Export_Invocation_-_GET/fhir/r4/Group/{logicalId}/$export_4). This seems fine for the API side of things, but not from an SFTP/Object Storage/Moveit (_shudder_) side of things.

**Comment by codyebberson** [comment]
Hi @jc00ke - Apologies for the very late reply.

I think you're looking for the FHIR `batch` API: https://hl7.org/fhir/r4/http.html#transaction

Specifically, the input to the FHIR `batch` is a FHIR `Bundle`, and the response will also be a FHIR `Bundle`.

The number of bundle entries in the response will be the same as the number of entries in the request.

Each `Bundle.entry` will have a `response` with helpful properties such as `status` and `outcome`.

See more: https://hl7.org/fhir/r4/bundle.html

At present, Medplum only supports synchronous FHIR `batch`.  We have a pull request open for asynchronous FHIR `batch` using `Prefer: respond-async`: https://github.com/medplum/medplum/pull/5308
